test/unit/math/mix/mat/fun/digamma_test.cpp

```diff
@@ -1,98 +1,7 @@
-#include <vector>
+#include <test/unit/math/test_ad.hpp>
 
-/**
- * This is the structure for testing vectorized digamma (defined in the
- * testing framework).
- */
-struct digamma_test {
-  /**
-   * Redefinition of function brought in from stan::math.  The reason
-   * to do this is that it wraps it up in this static template class.
-   *
-   * This is the version that's being tested.
-   *
-   * WARNING:  assumes that the scalar values for all instantiations
-   * (prim, rev, fwd, mix) ***have already been tested***.
-   *
-   * @tparam R Return type.
-   * @tparam T Argument type.
-   */
-  template <typename R, typename T>
-  static R apply(const T& x) {
-    using stan::math::digamma;
-    return digamma(x);
-  }
-
-  /**
-   * This defines the truth against which we're testing.
-   *
-   * Because this is *not an independent test*, this function just
-   * delegates to the actual function defined in stan::math.
-   *
-   * Redundant definition of function from stan::math to apply to an
-   * integer and return a double.
-   *
-   * This function delegates to apply(), defined above, directly.
-   *
-   * WARNING:  this is *not an independent test*.
-   */
-  static double apply_base(int x) { return apply<double>(x); }
-
-  /**
-   * This is the generic version of the integer version defined
-   * above.  For every other type, the return type is the same as the
-   * reference type.
-   *
-   * WARNING:  this is *not an independent test of the underlying function*.
-   */
-  template <typename T>
-  static T apply_base(const T& x) {
-    return apply<T>(x);
-  }
-
-  /**
-   * Return sequence of valid double-valued inputs.
-   */
-  static std::vector<double> valid_inputs() {
-    return test::math::vector_builder<double>()
-        .add(2.3)
-        .add(5.7)
-        .add(-10.2)
-        .add(-1.2)
-        .build();
-  }
-
-  /**
-   * Return sequence of invalid double-valued inputs.
-   */
-  static std::vector<double> invalid_inputs() {
-    return test::math::vector_builder<double>().build();
-  }
-
-  /**
-   * Return sequence of valid integer inputs.
-   */
-  static std::vector<int> int_valid_inputs() {
-    return test::math::vector_builder<int>().build();
-  }
-
-  /**
-   * Return sequence of invalid integer inputs.
-   */
-  static std::vector<int> int_invalid_inputs() {
-    return test::math::vector_builder<int>().add(-1).add(-12).add(-25).build();
-  }
-};
-
-INSTANTIATE_TYPED_TEST_CASE_P(, prim_scalar_unary_test, digamma_test);
-INSTANTIATE_TYPED_TEST_CASE_P(, rev_scalar_unary_test, digamma_test);
-INSTANTIATE_TYPED_TEST_CASE_P(, fwd_scalar_unary_test, digamma_test);
-INSTANTIATE_TYPED_TEST_CASE_P(, mix_scalar_unary_test, digamma_test);
+TEST(mathMixMatFun, digamma) {
+  auto f = [](const auto& x1) { return stan::math::digamma(x1); };
+  stan::test::expect_common_nonzero_unary_vectorized(f);
+  stan::test::expect_unary_vectorized(f, -25, -10.2, -1.2, -1, 2.3, 5.7);
```

Comment:
Yes.  It'll just instantiate a generic type `T` that just keeps getting passed along down to `expect_ad_helper`, where it gets passed directly to the functor `f`. I can add a test that makes it clear that this arduous path is actually working.

Reply:
Yeah would it be possible to add a few tests to make sure the right template deduction is happening in these automatic types and things aren't getting cast away?

Like make sure 1 is an int, 1.0 is a double, var stays var? Is this what you mean by the arduous path?

Reply:
Types get preserved because it's just templated arguments all the way down the call chain until the function being tested is invoked, and that's defined with a template parameter in order to match primitive and autodiff types.

I'll add some tests of the test framework to demonstrate from the user-callable test functions (I should split this into an internal helper file and what a developer's going to troll through to find the right test function).

The buck stops here, though---I don't intend to test the tests of the test framework. :-)

Reply:
> I can add a mock function that throws on integer values and check directly that it throws on int and not on double. Will that be enough?

Yeah. If I was writing tests, I would probably write some code to convince myself the correct thing was happening. Just too easy to promote to double.

If I know that this behavior is being checked for me I can rest easy.

Reply:
All tested now.  And the tests are tested (though you have to edit them to add in a test that fails to show that failure happens).  I didn't want to start thinking about how to start testing the Google test framework macros.

Reply:
Now onto part two.  

> So we expect all the error checking parts of functions to be checked in the prim checks? So that we can trust the prim implementations of the functions?

Yes.  These tests all assume the primitive versions return correct values and throw in the right places.  This framework does *not* test that.

> In that case, these functions were missing prim tests. Not sure I got all of them: acos, asin, atan, ceil, cos, fabs, floor, log10, sin, sinh, tanh

I'll add a bunch of automatic tests that make sure that `stan::math::F` returns same values and throws in same places as `std::F` for all the functions `F` we implement to the extent possible.  There's going to be trouble with `stan::math::log1p` because the behavior's been changed to not conform to `std::log1p`.   I might find other cases of that, in which case I obviously can't test vs `std::`.  I don't want this PR to degrade into testing or whole math lib!




ceil and floor, at least, should have tests. Want to add those in this pull or make another issue? I'd personally prefer to get them implemented (or found in case I missed them) now.

I need to test a function or two myself to understand the limits of this stuff. Do you see any difficulty testing more complicated stuff?

The matrix stuff, for instance? Like a cholesky, or matrix multiply, or something?